### ISSUE TITLE
[BUG]Fail to load checkpoints after zero3 initialize

### ISSUE THREAD

**Issue by mynewstart** [microsoft/DeepSpeed#4283]
**Describe the bug**
I can't load checkpoint when I use deepspeed.zero.Init(config_dict_or_path=ds_config,enabled=True)
**To Reproduce**
I want to continue pre-train the llama 30B model and load model checkpoint using deepspeed zero stage 3, the code like
```
 with deepspeed.zero.Init(config_dict_or_path=ds_config,
                             enabled=True,
                             mem_efficient_linear=False,
                             mpu=None):
        model_config = LlamaConfig.from_pretrained(Model_Type[args.model_type])
        model = LlamaForCausalLM.from_pretrained(Model_Type[args.model_type],config=model_config,torch_dtype=torch.float16)
optimizer_grouped_parameters = filter(lambda p: p.requires_grad, model.parameters())
custom_optimizer = FusedAdam(optimizer_grouped_parameters, lr=args.learning_rate,
                              betas=(0.9, 0.95))
custom_scheduler =  get_scheduler(
                name=args.lr_scheduler_type,
                optimizer=custom_optimizer,
                num_warmup_steps=int(args.num_warmup_steps//args.gradient_accumulation_steps),
                num_training_steps=int(args.total_training_steps//args.gradient_accumulation_steps),
            )
model_engine, _, _, _ = deepspeed.initialize(args=args,
                                                 model=model,
                                                 config = ds_config,
                                                 optimizer=custom_optimizer,
                                                 lr_scheduler= custom_scheduler,
                                                 dist_init_required=True,
                                                 )

```
I will met the below error information:
```
size mismatch for model.layers.0.self_attn.q_proj.weight: copying a param with shape torch.Size([6656, 6656]) from checkpoint, the shape in current model is torch.Size([0]).
size mismatch for model.layers.0.self_attn.k_proj.weight: copying a param with shape torch.Size([6656, 6656]) from checkpoint, the shape in current model is torch.Size([0]).
size mismatch for model.layers.0.self_attn.v_proj.weight: copying a param with shape torch.Size([6656, 6656]) from checkpoint, the shape in current model is torch.Size([0]).
size mismatch for model.layers.0.self_attn.o_proj.weight: copying a param with shape torch.Size([6656, 6656]) from checkpoint, the shape in current model is torch.Size([0]).
.....
```
**Expected behavior**
I know I can use with deepspeed.zero.Init(config_dict_or_path=ds_config) when I don't need to load model checkpoint, like 
```
 with deepspeed.zero.Init(config_dict_or_path=ds_config):
      model = LlamaForCausalLM(LlamaConfig(Model_Type[args.model_type]))
```
Or I can use HuggingFace Integration, use 
```
    if ds_config is not None and ds_config["zero_optimization"]["stage"] == 3:
        dschf = HfDeepSpeedConfig(ds_config)
    else:
        dschf = None
```
My question is Can I load checkpoint from deepspeed.zero.Init?

**ds_report output**
Please run `ds_report` to give us details about your setup.

**Screenshots**

**System info (please complete the following information):**
 - OS: Amazon linux2
 - GPU count and types A100 80G
 - Python version python=3.10
 - Any other relevant info about your setup

**Comment by tjruwase** [comment]
@mynewstart, [HfDeepSpeedConfig](https://huggingface.co/docs/transformers/main_classes/deepspeed#transformers.integrations.HfDeepSpeedConfig) was created for [HF+DeepSpeed](https://huggingface.co/docs/transformers/main_classes/deepspeed#nontrainer-deepspeed-integration) use cases like this. Is there any reason why you can't use it?

**Comment by mynewstart** [comment]
Hi @tjruwase, I am now using [HfDeepSpeedConfig](https://huggingface.co/docs/transformers/main_classes/deepspeed#transformers.integrations.HfDeepSpeedConfig), just want to know if we can use deepspeed.zero.Init() to load model checkpoint.
```
with deepspeed.zero.Init(config_dict_or_path=ds_config,
                             enabled=True,
                             mem_efficient_linear=False,
                             mpu=None):
```

**Comment by tjruwase** [comment]
@mynewstart, what I meant to say is that by using HfDeepSpeedConfig, you don't need to call `zero.Init()` in your script since the HF integration will automatically call [`zero.Init()`](https://github.com/huggingface/transformers/blob/04191ea1e6874fd9809f80f5567b3b9df911c8da/src/transformers/modeling_utils.py#L1177-L1184) during model initialization. This would load the checkpoint as well, as in this inference [example](https://github.com/huggingface/transformers-bloom-inference/blob/7bea3526d8270b4aeeefecc57d7d7d638e2bbe0e/bloom-inference-scripts/bloom-ds-zero-inference.py#L110-L118). Can you clarify why this behavior is insufficient for your case?

**Comment by mynewstart** [comment]
Hi @tjruwase Thanks for replying, turns out HFDeepSpeed call zero.Init()!

I previously attempted to train the 30B LLAMA model using zero3. When I directly load the mode without zero.Init(), each GPU loads all the weights, resulting in a training efficiency of 82 TFLOPS. However, when I use HfDeepSpeedConfig to load the model, each GPU only loads a partition of the weights, resulting in a training efficiency of 56 TFLOPS. I understand that the difference between the two only occurs during model loading, but both use zero3 for training. Therefore, I'm not sure why there is a difference in training efficiency.  So I suspected that using HfDeepSpeedConfig might be the cause, but based on your reply, it's not the reason.

**Comment by tjruwase** [comment]
That is a very significant performance loss. However, I think we need a new ticket for that problem. Do you mind opening a new ticket where you can share repro details? In particular, can please share log snippets that include latencies of forward, backward, optimizer step operations. 

Also, can you please close this ticket as I believe it is now resolved.

**Comment by mynewstart** [comment]
Sure, I will open a new ticket after I replay the training log. Thank you very much!

**Comment by nricciardi** [comment]
If you are really desperate and you could ignore all programming best practices (and in AI world this happens continuously), then you can simply load your model out of zero3 context. Therefore, use some environment variables to load the model in a dedicated file or outside your model wrap class and finally setup like this:
```
self.model = GLOBAL_MODEL
```
